NetKet/Machine/mps_diagonal.hpp

```diff
@@ -0,0 +1,471 @@
+
+  // Map from Hilbert states to MPS indices
+  std::map<double, int> confindex_;
+
+  const Hilbert &hilbert_;
+
+ public:
+  using StateType = T;
+  using LookupType = Lookup<T>;
+
+  // constructor as a machine
+  explicit MPSDiagonal(const Hilbert &hilbert, const json &pars)
+      : N_(hilbert.Size()), d_(hilbert.LocalSize()), hilbert_(hilbert) {
+    from_json(pars);
+  }
+
+  // Auxiliary function that defines the matrices
+  void Init() {
+    // Initialize parameters
+    std::vector<VectorType> pushback_vec;
+    VectorType init_mat = VectorType::Zero(D_);
+    npar_ = symperiod_ * d_ * D_;
+
+    for (int site = 0; site < symperiod_; site++) {
+      W_.push_back(pushback_vec);
+      for (int spin = 0; spin < d_; spin++) {
+        W_[site].push_back(init_mat);
+      }
+    }
+
+    // Machine creation messages
+
+    InfoMessage() << "Periodic diagonal MPS machine with " << N_
+                  << " sites created" << std::endl;
+    InfoMessage() << "Physical dimension d = " << d_
+                  << " and bond dimension D = " << D_ << std::endl;
+    if (symperiod_ < N_) {
+      InfoMessage() << "Translation invariance is used. Number of "
+                       "variational parameters is "
+                    << npar_ << " instead of " << npar_ * N_ / symperiod_
+                    << std::endl;
+    } else {
+      InfoMessage() << "Number of variational parameters is " << npar_
+                    << std::endl;
+    }
+
+    // Initialize map from Hilbert space states to MPS indices
+    auto localstates = hilbert_.LocalStates();
+    for (int i = 0; i < d_; i++) {
+      confindex_[localstates[i]] = i;
+    }
+  }
+
+  int Npar() const override { return npar_; }
+
+  VectorType GetParameters() override {
+    int k = 0;
+    VectorType pars(npar_);
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          pars(k) = W_[site][spin](i);
+          k++;
+        }
+      }
+    }
+    return pars;
+  }
+
+  void SetParameters(const VectorType &pars) override {
+    int k = 0;
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          W_[site][spin](i) = pars(k);
+          k++;
+        }
+      }
+    }
+  }
+
+  // Auxiliary function used for setting initial random parameters and adding
+  // identities in every matrix
+  inline void SetParametersIdentity(const VectorType &pars) {
+    int k = 0;
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          W_[site][spin](i) = T(1, 0) + pars(k);
+          k++;
+        }
+      }
+    }
+  }
+
+  void InitRandomPars(int seed, double sigma) override {
+    VectorType pars(npar_);
+
+    netket::RandomGaussian(pars, seed, sigma);
+    SetParametersIdentity(pars);
```

Comment:
Why is this necessary? Can't we initialize the matrices with random unitary matrices for example?